File: THCS/2022_2023/DakLak_2223/Bai3/BAI3.cpp
#include <iostream>
#include <vector>
#include <algorithm> // Để sử dụng std::lower_bound
using namespace std;

int main() {
    int N, Q;
    cin >> N >> Q;
    
    vector<int> a(N);
    
    // Đọc dãy số đã được sắp xếp
    for (int i = 0; i < N; ++i) {
        cin >> a[i];
    }

    // Xử lý mỗi truy vấn
    for (int i = 0; i < Q; ++i) {
        int L, R, S;
        cin >> L >> R >> S;

        // Chuyển đổi từ chỉ số 1-based sang 0-based
        L--; 
        R--;

        // Tìm vị trí số nhỏ nhất >= S trong đoạn a[L] đến a[R]
        auto it = lower_bound(a.begin() + L, a.begin() + R + 1, S);
        
        if (it != a.begin() + R + 1) { // Kiểm tra nếu tìm thấy số thỏa mãn
            cout << *it << endl;
        } else {
            cout << -1 << endl;
        }
    }

    return 0;
}

File: THCS/2022_2023/DakLak_2223/Bai3/BAI3.py
def main():
    import sys
    from bisect import bisect_left

    # Đọc số lượng phần tử và số truy vấn
    N, Q = map(int, input().split())
    
    # Đọc dãy số đã được sắp xếp
    a = list(map(int, input().split()))

    results = []

    # Xử lý mỗi truy vấn
    for _ in range(Q):
        L, R, S = map(int, input().split())
        # Chuyển đổi L, R từ 1-based sang 0-based
        L -= 1
        R -= 1

        # Tìm số nhỏ nhất >= S trong đoạn a[L] đến a[R]
        pos = bisect_left(a, S, L, R + 1)

        if pos <= R:  # Kiểm tra nếu pos nằm trong đoạn [L, R]
            results.append(a[pos])
        else:
            results.append(-1)

    # Xuất kết quả
    print('\n'.join(map(str, results)))

if __name__ == "__main__":
    main()
